components/main-chef-wrapper/main.go

```diff
@@ -0,0 +1,123 @@
+//
+// Copyright 2019 Chef Software, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+)
+
+func main() {
+	// No arguments provided, display usage
+	if len(os.Args) <= 1 {
+		usage()
+		os.Exit(0)
+	}
+	var (
+		subCommand = os.Args[1]
+		allArgs    = os.Args[2:]
+		cmd        *exec.Cmd
+	)
+
+	switch subCommand {
+	// 1) On the creation of new CLIs, for instance, a new CLI called chef-analyze that
+	// will provide some analyze capabilities, as a user, you can run the binary directly
+	// 'chef-analyze foo' or through our main Chef CLI wrapper 'chef analyze foo'.
+	// Both ways would be running the same underlying binary.
+	case "analyze":
+		cmd = exec.Command("chef-analyze", allArgs...)
+
+	// 2) Redirecting existing binaries to a single point for further improvements, this gives
+	// the potential to transition some commands at a more granular level too - for example
+	// 'knife search' can run a different a different binary than 'knife bootstrap'.
+	//
+	// TODO @afiune Propose this new behavior with the Team
+	//case "knife":
+	//cmd = exec.Command("knife", allArgs...)
+
+	// 3) On replacements of current functionality, for example, if we would like to improve
+	// the existing chef generate foo command, we would create a binary called chef-generate
+	// that users can run individually like chef-generate foo and then redirect the top-level
+	// CLI to run that same binary on any 'chef generate XYZ' execution.
+
+	// We want to pass every sub-command to the old 'chef' CLI binary that was renamed to
+	// 'chef-cli`, which is our default case.
+	default:
+		// When we land in the default case where we run the old 'chef' cli binary,
+		// we need to send the sub-command as well as all the arguments.
+		allArgs = append([]string{subCommand}, allArgs...)
+		cmd = exec.Command("chef-cli", allArgs...)
```

Comment:
In the final implementation, I don't think you want your wrapper depending on the system PATH for this.  Perhaps we can hard-code the expected omnibus paths as constants, inject them at build-time, or use a relative path based on the location of this file.

Reply:
This is a good enhancement @stevendanna - I will make sure to add an Issue to work on this!

Reply:
https://github.com/chef/chef-workstation/issues/562